### PR TITLE
feat(elixir-client): Add Move-In/Out Support for Subqueries in Elixir Client

### LINKED ISSUE #3672: Add Move-In/Out Support for Subqueries in Elixir Client
## Summary

Add support for tagged rows and move-out events in the Elixir client, enabling proper handling of shapes with subqueries in their WHERE clauses.

## Background

When a shape has a subquery (e.g., `WHERE parent_id IN (SELECT id FROM parent WHERE active = true)`), rows can dynamically enter or leave the shape based on changes in the dependency. The server implements this via:

- **Tags**: Each row gets tags explaining *why* it's in the shape (MD5 hash of the referenced value)
- **Move-out events**: When a dependency value is removed, a pattern-based event tells clients to remove matching rows
- **Snapshot-end**: Visibility boundaries for move-in query results

This feature was implemented in the TypeScript client via [TanStack/db#942](https://github.com/TanStack/db/pull/942) and the server-side in [electric-sql/electric#3427](https://github.com/electric-sql/electric/pull/3427).

### Change Messages with Tags

```json
{
  "key": "\"public\".\"child\"/\"1\"",
  "value": {"id": "1", "parent_id": "1", "name": "..."},
  "headers": {
    "operation": "insert",
    "tags": ["abc123def456..."]
  }
}
```

### Update with Tag Changes

```json
{
  "headers": {
    "operation": "update",
    "tags": ["xyz789..."],
    "removed_tags": ["abc123..."]
  }
}
```

### Move-Out Event

```json
{
  "headers": {
    "event": "move-out",
    "patterns": [{"pos": 0, "value": "abc123def456..."}]
  }
}
```

### Snapshot-End Control

```json
{
  "headers": {
    "control": "snapshot-end",
    "xmin": "100",
    "xmax": "200",
    "xip_list": ["150"]
  }
}
```

#### 1.1 Create `Electric.Client.TagIndex` module

Positional index for efficient move-out pattern matching.

**File**: `lib/electric/client/tag_index.ex`

| Function | Description |
|----------|-------------|
| `new/0` | Create empty index |
| `parse_tag/1` | Split tag string by `\|` delimiter |
| `add_tag/3` | Add tag to index for a row |
| `remove_tag/3` | Remove tag from index |
| `find_rows_matching_pattern/2` | O(1) lookup of rows by position+value |
| `tag_matches_pattern?/2` | Check if parsed tag matches pattern |

**Data structure**:
```elixir
%TagIndex{
  index: [%{value => MapSet.t(row_key)}],  # Array indexed by position
  tag_length: non_neg_integer() | nil
}
```

#### 1.2 Create `Electric.Client.MoveState` module

Track tag state for all rows in a shape.

**File**: `lib/electric/client/move_state.ex`

| Function | Description |
|----------|-------------|
| `new/0` | Create empty state |
| `add_tags_to_row/3` | Add tags to a row |
| `remove_tags_from_row/3` | Remove specific tags |
| `clear_row/2` | Remove all tags for a row (on delete) |
| `process_move_out_pattern/2` | Remove matching tags, return rows to delete |
| `reset/1` | Clear all state (on must-refetch) |

**Data structure**:
```elixir
%MoveState{
  row_tags: %{row_key => MapSet.t(tag)},
  tag_index: %TagIndex{},
  tag_cache: %{tag => parsed_tag}
}
```

---

#### 2.1 Update `Message.Headers`

Add tag fields to the headers struct:

```elixir
defstruct [
  # ... existing fields ...
  tags: [],           # List of move tags
  removed_tags: []    # Tags being removed (updates only)
]
```

#### 2.2 Add `Message.EventMessage`

New struct for event messages:

```elixir
defmodule EventMessage do
  defstruct [:event, :patterns, :handle, :request_timestamp]

  @type t :: %__MODULE__{
    event: :move_out,
    patterns: [%{pos: non_neg_integer(), value: String.t()}],
    handle: String.t(),
    request_timestamp: DateTime.t()
  }
end
```

#### 2.3 Update `Message.ControlMessage`

Add snapshot visibility fields for `snapshot-end`:

```elixir
defstruct [
  # ... existing fields ...
  xmin: nil,
  xmax: nil,
  xip_list: nil
]
```

#### 2.4 Update `Message.parse/3`

Add clause to parse event messages:

```elixir
def parse(%{"headers" => %{"event" => _}} = msg, handle, _) do
  [EventMessage.from_message(msg, handle)]
end
```

---

#### 3.1 Update Stream struct

```elixir
defstruct [
  # ... existing fields ...
  move_state: nil,
  buffered_move_outs: []
]
```

#### 3.2 Process tags on change messages

```elixir
defp process_change_tags(%ChangeMessage{} = msg, stream) do
  case msg.headers.operation do
    :delete -> MoveState.clear_row(stream.move_state, msg.key)
    _ ->
      stream.move_state
      |> MoveState.remove_tags_from_row(msg.key, msg.headers.removed_tags)
      |> MoveState.add_tags_to_row(msg.key, msg.headers.tags)
  end
end
```

#### 3.3 Handle move-out events

```elixir
defp handle_msg(%EventMessage{event: :move_out} = msg, stream) do
  if stream.up_to_date? do
    process_move_out(msg, stream)
  else
    # Buffer until initial sync completes
    {:cont, %{stream | buffered_move_outs: [msg | stream.buffered_move_outs]}}
  end
end
```

#### 3.4 Generate synthetic deletes

```elixir
defp process_move_out(%EventMessage{patterns: patterns}, stream) do
  {rows_to_delete, move_state} =
    Enum.reduce(patterns, {[], stream.move_state}, fn pattern, {dels, state} ->
      {new_dels, state} = MoveState.process_move_out_pattern(state, pattern)
      {new_dels ++ dels, state}
    end)

  # Generate synthetic delete messages
  delete_msgs = Enum.map(rows_to_delete, fn key ->
    %ChangeMessage{
      key: key,
      value: %{},
      headers: Headers.delete(handle: stream.shape_handle)
    }
  end)

  {:cont, buffer_messages(stream, delete_msgs)}
end
```

#### 3.5 Process buffered move-outs on up-to-date

```elixir
defp handle_msg(%ControlMessage{control: :up_to_date} = msg, stream) do
  stream = process_buffered_move_outs(stream)
  # ... existing up_to_date handling ...
end
```

---

#### Unit Tests

- [ ] `test/electric/client/tag_index_test.exs`
  - Tag parsing (simple, composite, escaped pipes)
  - Index add/remove operations
  - Pattern matching with wildcards

- [ ] `test/electric/client/move_state_test.exs`
  - Add/remove tags for rows
  - Process move-out patterns
  - Clear row on delete
  - Reset on must-refetch

#### Integration Tests

- [ ] `test/electric/client/move_integration_test.exs`
  - Receive tags on change messages
  - Process move-out events → synthetic deletes
  - Buffer move-outs during initial sync
  - Handle tag changes on updates
  - Handle must-refetch with move state

---

### Phase 5: Documentation

- [ ] Add `@moduledoc` to new modules
- [ ] Document tag format and pattern matching
- [ ] Add examples to README
- [ ] Document synthetic delete behavior

---

## Files Changed

| File | Action | Description |
|------|--------|-------------|
| `lib/electric/client/tag_index.ex` | **Create** | Positional tag index |
| `lib/electric/client/move_state.ex` | **Create** | Row tag state tracking |
| `lib/electric/client/message.ex` | Modify | Add EventMessage, update Headers |
| `lib/electric/client/stream.ex` | Modify | Process tags and move-outs |
| `test/electric/client/tag_index_test.exs` | **Create** | TagIndex unit tests |
| `test/electric/client/move_state_test.exs` | **Create** | MoveState unit tests |
| `test/electric/client/move_integration_test.exs` | **Create** | Integration tests |

---

## Acceptance Criteria

- [ ] Tags are parsed from change message headers
- [ ] Tags are tracked per row in `MoveState`
- [ ] Move-out events trigger removal of matching tags
- [ ] Rows with empty tag sets generate synthetic delete messages
- [ ] Move-out events are buffered during initial sync
- [ ] State is cleared on `must-refetch`
- [ ] Existing functionality (shapes without subqueries) is unchanged
- [ ] All tests pass

---

### Synthetic Deletes

Move-outs generate `%ChangeMessage{operation: :delete}` so consumers don't need special handling. The delete appears in the stream like any other delete.

### Buffering Strategy

Move-out events during initial sync are buffered and processed when `up-to-date` is received. This prevents deleting rows before they're inserted.

### Backward Compatibility

Tags are optional. Shapes without subqueries work exactly as before - the `move_state` is simply unused.

### Memory Considerations

Each tagged row adds entries to `row_tags` map and `tag_index`. For large datasets with many tags, memory usage scales linearly.

---

## References

- Server implementation: https://github.com/electric-sql/electric/pull/3427
- TypeScript client: https://github.com/TanStack/db/pull/942
- Key server files:
  - `sync-service/lib/electric/shapes/shape/subquery_moves.ex`
  - `sync-service/lib/electric/shapes/consumer/move_handling.ex`
  - `sync-service/lib/electric/shapes/consumer/move_ins.ex`

---

## Labels

`enhancement`, `elixir-client`, `subqueries`

### PR DESCRIPTION
**Pull request by robacourt**
Fixes #3672

  Summary

  - Implements tagged row and move-out event support in the Elixir client for shapes with subqueries in WHERE clauses
  - Introduces end-to-end integration tests that validate the full client/server behavior
  - Adds StreamConsumer test helper for expressive, human-readable test assertions

  The Problem

  When a shape contains a subquery like WHERE parent_id IN (SELECT id FROM parent WHERE active = true), rows can dynamically enter or exit the shape based on changes to dependency tables. For example, deactivating a parent row should cause all its children to be removed from the shape (move-out), even though the child rows themselves weren't modified.

  The Solution

  The Elixir client now:
  1. Tracks tags on change messages (MD5 hashes indicating why rows belong to the shape)
  2. Processes move-out control messages from the server
  3. Generates synthetic delete messages when move-outs occur
  4. Maintains a tag index to correctly handle edge cases (e.g., a row changing parents)

  Testing Approach

  A key aspect of this PR is the integration test design. Rather than relying solely on unit tests with mocks, these tests spin up the full Electric stack and exercise real scenarios:

```elixir
  test "receives move-out when parent is deactivated", %{client: client, shape: shape, db_conn: db_conn} do
    stream = Client.stream(client, shape, live: true)

    with_consumer stream do
      # Wait for initial snapshot
      assert_insert(consumer, %{"id" => "child-1"})
      assert_up_to_date(consumer)

      # Deactivate the parent - this should trigger a move-out
      Postgrex.query!(db_conn, "UPDATE parent SET active = false WHERE id = 'parent-1'", [])

      # Should receive a synthetic delete for child-1
      assert_delete(consumer, %{"id" => "child-1"})
    end
  end
```

  Benefits of this approach:
  - Readable as specifications: Tests describe expected behavior in plain terms that humans can verify
  - Oracle for AI development: Tests serve as ground truth that AI tools can use to drive implementation and verify correctness
  - Easier review: Reviewers can easily understand that the system is doing what it's supposed to at a high level without diving into implementation details
  - Confidence: Tests exercise the actual code paths through the full system, not just isolated units

  Changes

  | Area                       | Changes                                                                  |
  |----------------------------|--------------------------------------------------------------------------|
  | electric/client/stream.ex  | Move-out processing, tag index tracking, synthetic delete generation     |
  | electric/client/message.ex | Tag and event message parsing                                            |
  | Integration tests          | 8 end-to-end scenarios covering move-outs, tag updates, resume, etc.     |
  | StreamConsumer helper      | Expressive test DSL with assert_insert, assert_delete, assert_up_to_date |

  Test Plan

  - Unit tests for message parsing and tag operations
  - Integration tests for move-out scenarios
  - Resume behavior after move-out
  - Edge case: changing parent doesn't generate spurious deletes
  - Backward compatibility with untagged shapes

  Additional changes separated from the main feature:

  Bug Fixes (candidates for backporting to main)

  1. Fix empty error array handling (stream.ex)
    - Renamed unwrap → unwrap_error and added handling for empty error arrays from the server
    - Previously [] would be returned as the error message; now returns "Unknown error"
  2. Fix resume to use live requests for long-polling (stream.ex)
    - When resuming a stream, the client was making non-live requests because up_to_date? defaulted to false
    - Since a ResumeMessage is only created when already up-to-date, resuming should set up_to_date?: true so requests use long-polling

  Infrastructure (supporting integration tests)

  3. Increase Postgres max_connections
    - Set max_connections = 200 in dev configs and CI to support integration tests with multiple concurrent connections
  4. Reduce default connection pool size
    - Changed default pool_size from 2 to 1 in test setup
    - Tests needing more connections use explicit @tag connection_opt_overrides: [pool_size: 2]
  5. Guard Ecto-dependent code
    - Wrapped ArrayDecoder module in Code.ensure_loaded?(Ecto) to prevent compile errors when Ecto is unavailable

  Note: The commit history has been carefully organized to separate bug fixes, infrastructure changes, and the core feature implementation (by claude!). The two bug fixes in commits 1-2 are independent improvements discovered during development that could be backported to main.

## Release Notes

* **New Features**
  * Added move-out support for subqueries with automatic synthetic delete generation when dependencies become inactive
  * Enhanced message headers with tag tracking and removal support for improved message filtering
  * Improved stream state management for dependency tracking and tag-based message organization

* **Tests**
  * Added comprehensive integration tests for streaming and move-out scenarios

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/elixir-client/lib/electric/client/message.ex [hunk 1]
```diff
@@ -168,15 +182,58 @@ defmodule Electric.Client.Message do
...
+    """
+
+    defstruct [:patterns, :handle, :request_timestamp]
+
+    @type pattern :: %{pos: non_neg_integer(), value: String.t()}
+    @type t :: %__MODULE__{
+            patterns: [pattern()],
+            handle: Client.shape_handle(),
+            request_timestamp: DateTime.t()
+          }
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/lib/electric/client/message.ex:220]: It was set, just not in the parser, similar to other messages. But I agree it's confusing so I've moved it so it is now set in the parser.

File: packages/elixir-client/lib/electric/client/ecto_adapter/array_decoder.ex [hunk 2]
```diff
@@ -1,92 +1,94 @@
...
-    do: decode_quoted_elem(rest, [acc | <<c::utf8>>], state)
+    defp decode_quoted_elem(<<c::utf8, rest::bitstring>>, acc, state),
+      do: decode_quoted_elem(rest, [acc | <<c::utf8>>], state)
 
-  defp decode_quoted_elem("", _acc, {_cast_fun, type, source}) do
-    raise Ecto.CastError, type: {:array, type}, value: source
-  end
+    defp decode_quoted_elem("", _acc, {_cast_fun, type, source}) do
+      raise Ecto.CastError, type: {:array, type}, value: source
+    end
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/lib/electric/client/ecto_adapter/array_decoder.ex:75]: This was not a change introduced by this PR

File: packages/sync-service/test/test_helper.exs [hunk 3]
```diff
@@ -7,6 +7,12 @@
 ExUnit.configure(formatters: [JUnitFormatter, ExUnit.CLIFormatter])
 ExUnit.start(assert_receive_timeout: 400, exclude: [:slow], capture_log: true)
 
+# Start electric_client application directly, bypassing OTP's dependency resolution.
+# This avoids a circular dependency: electric_client has :electric as an optional dep,
+# which gets added to its applications list when compiled in sync-service context,
+# causing a deadlock when OTP tries to start both applications.
+{:ok, _} = Electric.Client.Application.start(:normal, [])
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/sync-service/test/test_helper.exs:14]: how would that happen?

File: packages/elixir-client/lib/electric/client/ecto_adapter/array_decoder.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/lib/electric/client/ecto_adapter/array_decoder.ex:75]: sure

File: packages/elixir-client/test/electric/client/message_test.exs [hunk 4]
```diff
@@ -29,4 +29,140 @@ defmodule Electric.Client.MessageTest do
...
+            %{"pos" => 2, "value" => "tag-3"}
+          ]
+        }
+      }
+
+      assert [%MoveOutMessage{patterns: patterns}] = Message.parse(msg, "handle", & &1)
+      assert length(patterns) == 3
+      assert Enum.at(patterns, 0) == %{pos: 0, value: "tag-1"}
+      assert Enum.at(patterns, 1) == %{pos: 1, value: "tag-2"}
+      assert Enum.at(patterns, 2) == %{pos: 2, value: "tag-3"}
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/test/electric/client/message_test.exs]: you could even do that as a single assert

[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/test/electric/client/message_test.exs]: are you not going that as a committable suggestion? :)

File: packages/elixir-client/lib/electric/client/stream.ex [hunk 5]
```diff
@@ -221,9 +230,28 @@ defmodule Electric.Client.Stream do
...
+
+    # Add synthetic deletes to the buffer
+    buffer =
+      Enum.reduce(synthetic_deletes, stream.buffer, fn delete_msg, buf ->
+        :queue.in(delete_msg, buf)
+      end)
+
+    {:cont,
+     %{stream | buffer: buffer, tag_to_keys: updated_tag_to_keys, key_data: updated_key_data}}
+  end
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/lib/electric/client/stream.ex]: I'm not sure you could reliably make the assert in the client. You could keep state of whether an up_to_date had been seen, but plausibly you could have snapshot, move-out, up to date which would still be fine but would then break the assertion.

File: packages/sync-service/test/integration/subquery_move_out_test.exs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        update_msg = assert_update(consumer, %{"id" => "child-1"})
+
+        # The headers should include both removed_tags and new tags
+        new_tags = Map.get(update_msg.headers, :tags, [])
+        removed_tags = Map.get(update_msg.headers, :removed_tags, [])
+
+        # Either we have explicit removed_tags, or tags have changed
+        assert new_tags != initial_tags or length(removed_tags) > 0
+      end
+    end
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/sync-service/test/integration/subquery_move_out_test.exs:218]: I'm fine with it as is

File: packages/sync-service/test/integration/subquery_move_out_test.exs [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+          assert_up_to_date(consumer)
+          assert_resume(consumer)
+        end
+
+      # Now deactivate the parent while "disconnected"
+      # This should trigger a move-out on the server side
+      Postgrex.query!(db_conn, "UPDATE parent SET active = false WHERE id = 'parent-1'", [])
+
+      # Give the server time to process the change
+      Process.sleep(100)
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/sync-service/test/integration/subquery_move_out_test.exs]: good spot! this has been fixed

File: packages/elixir-client/lib/electric/client/message.ex [hunk 8]
```diff
@@ -168,30 +187,108 @@ defmodule Electric.Client.Message do
...
+        _value_mapper_fun,
+        request_timestamp
+      ) do
+    [MoveOutMessage.from_message(msg, shape_handle, request_timestamp)]
   end
 
-  def parse("", _handle, _value_mapper_fun) do
+  def parse("", _handle, _value_mapper_fun, _request_timestamp) do
     []
   end
```
[the remark this answers is not in the document]
Comment (robacourt) [reply packages/elixir-client/lib/electric/client/message.ex:293]: This has nothing to do with this PR, and I think this is fine for now

File: .github/workflows/sync_service_tests.yml [hunk 9]
```diff
@@ -42,6 +42,7 @@ jobs:
           --health-retries 5
         ports:
           - 54321:5432
+        command: postgres -c max_connections=200
```
[the remark this answers is not in the document]
Comment (robacourt) [reply .github/workflows/sync_service_tests.yml:45]: that claim is false. The command key is actually supported for GitHub Actions service containers.

[the remark this answers is not in the document]
Comment (robacourt) [reply .github/workflows/sync_service_tests.yml:45]: not true

**Review by icehaunter — APPROVED** [review]
